File: shared/contracts/python/governance_decision.py
"""
GovernanceDecision contract — Python / Pydantic.
Represents the outcome of evaluating a request against governance policy.
Every AI action produces exactly one GovernanceDecision before proceeding.
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum

from pydantic import BaseModel


class GovernanceOutcome(str, Enum):
    ALLOW = "ALLOW"
    DENY = "DENY"
    ESCALATE = "ESCALATE"


class GovernanceDenyReason(str, Enum):
    POLICY_VIOLATION = "POLICY_VIOLATION"
    INSUFFICIENT_PERMISSION = "INSUFFICIENT_PERMISSION"
    OUT_OF_SCOPE = "OUT_OF_SCOPE"
    PII_EXPOSURE_RISK = "PII_EXPOSURE_RISK"
    HIGH_RISK_ACTION_BLOCKED = "HIGH_RISK_ACTION_BLOCKED"


class GovernanceEscalateReason(str, Enum):
    HUMAN_APPROVAL_REQUIRED = "HUMAN_APPROVAL_REQUIRED"
    HIGH_RESERVE_AMOUNT = "HIGH_RESERVE_AMOUNT"
    LITIGATION_FLAG = "LITIGATION_FLAG"
    LOW_MODEL_CONFIDENCE = "LOW_MODEL_CONFIDENCE"
    SENSITIVE_CONTENT_DETECTED = "SENSITIVE_CONTENT_DETECTED"


class PolicyEvaluation(BaseModel):
    policy_id: str
    policy_version: str
    rule_id: str
    matched: bool
    outcome: GovernanceOutcome
    reason: str


class GovernanceDecision(BaseModel):
    decision_id: str
    evaluated_at: datetime
    task_type: str
    claim_id: str | None = None

    # Final outcome — strictest rule wins (DENY > ESCALATE > ALLOW)
    outcome: GovernanceOutcome

    # Human-readable explanation always present
    reason: str

    deny_reason: GovernanceDenyReason | None = None
    escalate_reason: GovernanceEscalateReason | None = None

    # Individual policy evaluations that led to this decision
    policy_evaluations: list[PolicyEvaluation] = []

    policy_set_id: str
    policy_set_version: str

File: shared/contracts/typescript/governanceDecision.ts
// GovernanceDecision contract
// Represents the outcome of evaluating a request against governance policy.
// Every AI action produces exactly one GovernanceDecision before proceeding.

export enum GovernanceOutcome {
  Allow = "ALLOW",
  Deny = "DENY",
  Escalate = "ESCALATE",
}

export enum GovernanceDenyReason {
  PolicyViolation = "POLICY_VIOLATION",
  InsufficientPermission = "INSUFFICIENT_PERMISSION",
  OutOfScope = "OUT_OF_SCOPE",
  PiiExposureRisk = "PII_EXPOSURE_RISK",
  HighRiskActionBlocked = "HIGH_RISK_ACTION_BLOCKED",
}

export enum GovernanceEscalateReason {
  HumanApprovalRequired = "HUMAN_APPROVAL_REQUIRED",
  HighReserveAmount = "HIGH_RESERVE_AMOUNT",
  LitigationFlag = "LITIGATION_FLAG",
  LowModelConfidence = "LOW_MODEL_CONFIDENCE",
  SensitiveContentDetected = "SENSITIVE_CONTENT_DETECTED",
}

export interface PolicyEvaluation {
  policyId: string;
  policyVersion: string;
  ruleId: string;
  matched: boolean;
  outcome: GovernanceOutcome;
  reason: string;
}

export interface GovernanceDecision {
  decisionId: string;
  evaluatedAt: string;             // ISO 8601
  taskType: string;                // e.g. "claim_summary", "draft_note"
  claimId: string | null;

  // Final outcome — strictest rule wins (DENY > ESCALATE > ALLOW)
  outcome: GovernanceOutcome;

  // Human-readable explanation always present
  reason: string;

  // Populated when outcome = DENY
  denyReason: GovernanceDenyReason | null;

  // Populated when outcome = ESCALATE
  escalateReason: GovernanceEscalateReason | null;

  // Individual policy evaluations that led to this decision
  policyEvaluations: PolicyEvaluation[];

  // Which policy set was active
  policySetId: string;
  policySetVersion: string;
}
